presto-main/src/main/java/com/facebook/presto/execution/scheduler/nodeSelection/SimpleNodeSelector.java

```diff
@@ -133,13 +137,34 @@ public SplitPlacementResult computeAssignments(Set<Split> splits, List<RemoteTas
         OptionalInt preferredNodeCount = OptionalInt.empty();
         for (Split split : splits) {
             List<InternalNode> candidateNodes;
+            boolean isCacheable = true;
             switch (split.getNodeSelectionStrategy()) {
                 case HARD_AFFINITY:
                     candidateNodes = selectExactNodes(nodeMap, split.getPreferredNodes(sortedCandidates), includeCoordinator);
                     preferredNodeCount = OptionalInt.of(candidateNodes.size());
                     break;
                 case SOFT_AFFINITY:
-                    candidateNodes = selectExactNodes(nodeMap, split.getPreferredNodes(sortedCandidates), includeCoordinator);
+                    List<HostAddress> preferredNodes;
```

Comment:
By using the nodemap as the only structure for nodes information, along with state as well, we can rely on most of the existing implementation. We only need to check at the last moment, if the candidate is not available, we just move to next candidate, and then to the least busy worker.